### PR TITLE
[sw, base] Add a macro for creating addressable code labels

### PR DESCRIPTION
**Pull request by mcy**
This macro is intended for writing chip-level tests that want to check
that some kind of exception-ey thing happened in a range of program
counters. It is a maximally compiler-portable implementation that won't
be optimization-hostile like GCC's label references (&&label), whose
support in LLVM is... not something we should rely on.

This macro is mostly made of caveats, but it has a narrow use
regardless.

Signed-off-by: Miguel Young de la Sota <mcyoung@google.com>

### CODE REVIEW

**Comment by matutem** [comment]
Cool, I'll try this prior to reviewing.

File: sw/device/lib/base/macros.h [hunk 1]
```diff
@@ -162,4 +162,51 @@
...
+ * producing a garbage value on the host side.
+ *
+ * The symbol reference `kName_` will only be scoped to the current block
+ * in a function, but it can be redeclared with
+ *
+ * ```
+ * extern const char kName_[];
+ * ```
+ *
+ * if needed elsewhere (although this should be avoided for readbility's sake).
```
Comment (luismarques) [inline sw/device/lib/base/macros.h:200]: nit: typo (readbility)

**Comment by luismarques** [comment]
Can you please add an example showcasing a typical usage? While you already did write a nice overview, it was mostly written from the point of view of the caveats. I had some thoughts about the mechanism but I'm not sure they apply to intended usage scenarios.

**Comment by mcy** [comment]
Luis: I've added an example, which also illustrates the function call caveat of trying to get a PC range. I don't think we can particularly do better, unless LLVM happens to have a way for ensuring no call instructions are generated in the backend.

File: sw/device/lib/base/macros.h [hunk 2]
```diff
@@ -162,4 +162,82 @@
...
+ *
+ * if needed elsewhere (although this should be avoided for readability's sake).
+ * The name of the constant is global to the program, so it should be given a
+ * name that is not going to collide (such as `kMyFunctionBeforeLoop`).
+ *
+ * # Example Uses
+ */
+#define OT_ADDRESSABLE_LABEL(kName_)                                         \
+  extern const char kName_[];                                                \
+  asm volatile(".global " #kName_ "; " #kName_ ":;");                        \
```
Comment (luismarques) [inline sw/device/lib/base/macros.h:237]: Shouldn't this be a `.local` directive instead?

File: sw/device/lib/base/macros.h [hunk 3]
```diff
@@ -162,4 +162,82 @@
...
+ *
+ * ```
+ * OT_ADDRESSABLE_LABEL(kIrqWaitStart);
+ * enable_interrupts();
+ * while(!irq_happened) {
+ *   wait_for_interrupt();
+ * }
+ * OT_ADDRESSABLE_LABEL(kIrqWaitEnd);
+ *
+ * CHECK(irq_pc >= &kIrqWaitStart && irq_pc <= &IrqWaitEnd);
```
Comment (luismarques) [inline sw/device/lib/base/macros.h:188]: I think the `<=` should be `<`?

File: sw/device/lib/base/macros.h [hunk 4]
```diff
@@ -162,4 +162,82 @@
...
+ * This has the same semantics as a `volatile` inline assembly block: it
+ * may be reordered with respect to all operations except other volatile
+ * operations (i.e. volatile assembly and volatile read/write, such as
+ * MMIO). For example, in the following code, `kBefore` and `kAfter` can
+ * wind up having the same address:
+ *
+ * ```
+ * OT_ADDRESSABLE_LABEL(kBefore);
+ * x += 5;
+ * OT_ADDRESSABLE_LABEL(kAfter);
```
Comment (luismarques) [inline sw/device/lib/base/macros.h:207]: The semantics of this mechanism are sound when we access a volatile variable between the `kBefore` and `kAfter`. While that is kind of implied in this caveat and in the interrupt example, it is never explicitly stated as a requirement. Maybe consider making that point a bit more explicit?

**Review by luismarques — COMMENTED** [review]
This is looking good. There are some minor points that you might want to address (see my inline comments), but overall it looks ready to merge.

File: sw/device/lib/base/macros.h (hunk 2 above)
Comment (mcy) [reply sw/device/lib/base/macros.h:237]: Neat, I did not know there was a way to specify that. It seems elf-only, so if someone's mac build breaks we'll sort out a workaround for them when the time comes.

**Review by matutem — APPROVED** [review]
I have not tested this yet, but no point delaying this PR.
LGTM

**Pull request merged**